### PR TITLE
Various performance optimizations and bug fixes

### PR DESCRIPTION
**Pull request by xichen2020**
cc @robskillington @cw9 @prateek @jeromefroe 

This PR contains various performance optimizations for the aggregation tier. Specifically, it includes
* Adding an option to provide a nil sample pool, which effectively disables sample pooling. The sample pool is used to provide sample objects for every timer value, which is on the critical path and is heavily contended. Profile shows getting values from the pool and putting values into the pool accounts for >30% of the CPU due to heavy contention. Disabling sample pooling significantly reduced write p99 (4ms -> 1ms) and flush p99 (1.5s -> 200ms).
* Adding an option to periodically flush streams at write time, therefore reducing processing load during global flushing. This has significantly improved flush p99 for aggregation servers that have large batches of timer values. In one case, we saw 5s -> 150ms reduction w.r.t. p99 flush latency.
* Replace math.Pow(x, 2) with x * x. This used to account for 4% of the CPU due to math.Pow(x, 2) being 100x more expensive than the plain multiplication. 
* Preallocate enough buffer for the buffered encoder. The buffered encode is used to hold the aggregated metrics. Previous it starts with an empty buffer and gradually grows to the maximum flush size, which requires unnecessary copy and reallocations.

It also includes more instrumentation and bug fixes. Specifically, it includes:
* Fixing a memory leak in timer creation where the underlying stream is not reset, causing the stream (and the float slices and the samples) to not be properly returned to pool.
* Adding logging to record the remote address when a decode error happens. This has allowed us to determine the primary cause of the decode errors is due to misconfigured hosts in the wrong datacenter.

### CODE REVIEW

File: aggregation/quantile/cm/options.go [hunk 1]
```diff
@@ -61,9 +62,10 @@ type options struct {
 // NewOptions creates a new options
```
Comment (jeromefroe) [inline aggregation/quantile/cm/options.go]: Missing a period at the end of the comment

File: aggregation/quantile/cm/stream.go [hunk 2]
```diff
@@ -34,17 +34,26 @@ var (
 	nan = math.NaN()
 )
 
+// acquireSampleFn acquires a new sample.
+type acquireSampleFn func() *Sample
+
+// releaseSampleFn releases a sample.
+type releaseSampleFn func(*Sample)
+
 // stream represents a data stream
```
Comment (jeromefroe) [inline aggregation/quantile/cm/stream.go]: nit: missing a period

File: aggregation/quantile/cm/options.go (hunk 1 above)
Comment (xichen2020) [inline aggregation/quantile/cm/options.go]: Yeah I'm gonna put up a separate PR that does a blanket sweep across the entire m3aggregator repo looking for comments missing periods.

File: aggregation/quantile/cm/stream.go (hunk 2 above)
Comment (xichen2020) [inline aggregation/quantile/cm/stream.go]: See above

File: aggregation/quantile/cm/stream.go [hunk 3]
```diff
@@ -311,7 +345,9 @@ func (s *stream) addToMinHeap(heap *minHeap, value float64) {
 	// and return the current heap to the pool
 	if len(curr) == cap(curr) {
 		newHeap := s.floatsPool.Get(2 * len(curr))
-		newHeap = append(newHeap, curr...)
+		for i := 0; i < len(curr); i++ {
```
Comment (jeromefroe) [inline aggregation/quantile/cm/stream.go]: Might be cheaper to use `copy` here, i.e. `copy(newHeap, curr)`.

File: aggregation/quantile/tdigest/tdigest.go [hunk 4]
```diff
@@ -297,7 +297,9 @@ func (d *tDigest) upperBound(index int) float64 {
 func (d *tDigest) appendCentroid(centroids []Centroid, c Centroid) []Centroid {
 	if len(centroids) == cap(centroids) {
 		newCentroids := d.centroidsPool.Get(2 * len(centroids))
-		newCentroids = append(newCentroids, centroids...)
+		for i := 0; i < len(centroids); i++ {
```
Comment (jeromefroe) [inline aggregation/quantile/tdigest/tdigest.go]: Same as above, maybe we want to use `copy` here.

File: services/m3aggregator/config/aggregator.go [hunk 5]
```diff
@@ -46,6 +46,10 @@ import (
 	"github.com/spaolacci/murmur3"
 )
 
+const (
+	initialBufferSizeGrowthFactor = 2
```
Comment (jeromefroe) [inline services/m3aggregator/config/aggregator.go]: Do we want to make this twice the flush size? Seems like that might be overkill. As a possible alternative we could make this a function call so we could use a tighter bound. For example:

```
func (c *AggregatorConfiguration) getBufferSize(n int) int {
    // Buffer size should be 1.5 times the maximum flush size to account for overflow.
    return c.MaxFlushSize * 3 * 2
}

```

File: services/m3aggregator/config/aggregator.go [hunk 6]
```diff
@@ -282,11 +286,16 @@ func (c *AggregatorConfiguration) NewAggregatorOptions(
 	entryPool.Init(func() *aggregator.Entry { return aggregator.NewEntry(nil, opts) })
 
 	// Set buffered encoder pool.
+	// NB(xichen): we preallocate a bit over the maximum flush size as a safety measure
```
Comment (jeromefroe) [inline services/m3aggregator/config/aggregator.go]: Perhaps move this line to before the line where we determine the `intialBufferSize` just so its a littler clearer that that is the line its referring to?

File: aggregation/quantile/cm/stream.go (hunk 3 above)
Comment (xichen2020) [inline aggregation/quantile/cm/stream.go]: The newly allocated slice is a zero-length slice with bigger capacity, don't think you can use `copy` here.

File: aggregation/quantile/tdigest/tdigest.go (hunk 4 above)
Comment (xichen2020) [inline aggregation/quantile/tdigest/tdigest.go]: The newly allocated slice is a zero-length slice with bigger capacity, don't think you can use `copy` here.

File: services/m3aggregator/config/aggregator.go (hunk 5 above)
Comment (xichen2020) [inline services/m3aggregator/config/aggregator.go]: Why 1.5 times? I think it roughly doubles the capacity if you write past the pre-allocated capacity.

File: services/m3aggregator/config/aggregator.go (hunk 6 above)
Comment (xichen2020) [inline services/m3aggregator/config/aggregator.go]: Sure thing

File: aggregation/quantile/cm/stream.go (hunk 3 above)
Comment (jeromefroe) [inline aggregation/quantile/cm/stream.go]: Aggh, yea, if the slice we get back has zero length then `copy` won't work. Any reason to not use `append(newHeap, curr...)` then? It seems like the compiler should be able to optimize that into a single copy under the hood.

File: services/m3aggregator/config/aggregator.go (hunk 5 above)
Comment (jeromefroe) [inline services/m3aggregator/config/aggregator.go]: It doesn't have to be 1.5 times, that was just for illustration. I'm just thinking that if the buffer is, for example, 10 kB, it might make more sense to allocate a buffer that is 12 kB instead of 20 kB, since the the overflow will likely be less than 2kB.

Comment (xichen2020) [inline services/m3aggregator/config/aggregator.go]: Sure, I can make it a bit more conservative, e.g., 1.5x. The number of buffers is actually pretty small (8096) so even allocating 2x we are allocating 8MB more (25MB vs. 17MB)

File: aggregation/quantile/cm/stream.go (hunk 3 above)
Comment (xichen2020) [inline aggregation/quantile/cm/stream.go]: I don't trust the compiler to not allocate a new slice so I'd rather do it myself =p

Comment (jeromefroe) [inline aggregation/quantile/cm/stream.go]: Fair enough :)

File: services/m3aggregator/config/aggregator.go (hunk 5 above)
Comment (jeromefroe) [inline services/m3aggregator/config/aggregator.go]: Hmm, in that case, perhaps we should just go with the simple approach of using 2x. I was thinking that the number of buffers would be closer to the number of connections (~15-20K) in which case the savings might be more significant, especially if we ever decided to increase the size of the buffers themselves.

File: aggregation/quantile/cm/stream.go (hunk 3 above)
Comment (cw9) [inline aggregation/quantile/cm/stream.go]: I think it def allocates a new slice when you pass in curr...

File: aggregator/list.go [hunk 7]
```diff
@@ -110,7 +112,7 @@ func newMetricList(resolution time.Duration, opts Options) *metricList {
 		resolution:    resolution,
 		flushInterval: flushInterval,
 		aggregations:  list.New(),
-		encoder:       msgpack.NewAggregatedEncoder(encoderPool.Get()),
+		encoder:       msgpack.NewAggregatedEncoder(encoder),
```
Comment (cw9) [inline aggregator/list.go]: can we do the encoder.Reset() in encoderPool.Get()? so you don't have to call this every time.

File: server/msgpack/server.go [hunk 8]
```diff
@@ -205,7 +205,14 @@ func (s *server) handleConnection(conn net.Conn) {
 	// If there is an error during decoding, it's likely due to a broken connection
 	// and therefore we ignore the EOF error
 	if err := it.Err(); err != nil && err != io.EOF {
-		s.log.Errorf("decode error: %v", err)
+		remoteAddress := "<unknown>"
```
Comment (cw9) [inline server/msgpack/server.go]: nit: make a const for "<unknown>"?

File: aggregation/quantile/cm/stream.go (hunk 3 above)
Comment (jeromefroe) [inline aggregation/quantile/cm/stream.go]: I don't think it allocates a new slice for `curr..` but creates a subslice for it instead. I wrote some [quick benchmarks](https://github.com/jeromefroe/golang_benchmarks#append) and the `curr..` approach was actually much faster and didn't cause any allocations.

Comment (xichen2020) [inline aggregation/quantile/cm/stream.go]: Yeah it doesn't allocate a new slice (see https://play.golang.org/p/MUxy9zCY9W), which is somewhat surprising in itself.

Comment (cw9) [inline aggregation/quantile/cm/stream.go]: hmm, in this case we should probably use the curr.. since the test shows there is no allocation

Comment (xichen2020) [inline aggregation/quantile/cm/stream.go]: Sounds good, will revert the change

File: aggregation/quantile/tdigest/tdigest.go (hunk 4 above)
Comment (xichen2020) [inline aggregation/quantile/tdigest/tdigest.go]: Will revert the change

File: aggregator/list.go (hunk 7 above)
Comment (xichen2020) [inline aggregator/list.go]: It's somewhat conventional to do `object := pool.Get(); object.Reset()` before using the pooled object, this is also how we use pooled objects in m3db.

File: server/msgpack/server.go (hunk 8 above)
Comment (xichen2020) [inline server/msgpack/server.go]: Sure thing
